src/windows/main_window.py

```diff
@@ -1227,13 +1227,35 @@ def actionAddMarker_trigger(self, event):
         marker.data = {"position": position, "icon": "blue.png"}
         marker.save()
 
-    def actionPreviousMarker_trigger(self, event):
-        log.info("actionPreviousMarker_trigger")
 
-        # Calculate current position (in seconds)
-        fps = get_app().project.get("fps")
-        fps_float = float(fps["num"]) / float(fps["den"])
-        current_position = (self.preview_thread.current_frame - 1) / fps_float
+    def findAllMarkerPositions(self):
+        """Build and return a list of all seekable locations for the currently-selected timeline elements"""
+
+        def getTimelineObjectPositions(obj):
+            """ Add boundaries & all keyframes of a timeline object (clip, transition...) to all_marker_positions """
+            positions = []
+
+            fps = get_app().project.get("fps")
+            fps_float = float(fps["num"]) / float(fps["den"])
+
+            clip_start_time=obj.data["position"]
+            clip_stop_time=obj.data["position"] + (obj.data["end"] - obj.data["start"])
+
+            # add clip boundaries
+            all_marker_positions.append(clip_start_time)
```

Comment:
```suggestion
            positions.append(clip_start_time)
```

Better to make function `getTimelineObjectPositions(obj)` outside the method and class (before _MainWindow_). Just illogical - you modifying some unknown object and later returning other object, and both actions are in one function.

Reply:
I do agree about changing `all_marker_positions` to `positions` in `getTimelineObjectPositions(obj)`, however I'm not sure about the placement of the function. Unless it is used by something else than `findAllMarkerPositions()`, I think it is cleanest to leave it inside. And otherwise, maybe finding another place than `main_window.py`, if it is not inside the `MainWindow` class

Reply:
Nested functions is hard to read. This usually introduces mistakes in the code.

Performance hit:
https://stackoverflow.com/questions/14055843/what-is-the-performance-overhead-of-nested-functions

Reply:
^ Not a meaningful one. As the conclusion to that answer says,
> Remember that in this trivial case creating and calling functions is all the interpreter does. In any real usage the difference will be much less.

The difference, in terms of OpenShot's total CPU cycles, will be negligible at worst, most likely undetectable.

Locally-declared worker functions are an accepted pattern in high-level languages like Python and JavaScript. Heck, they're even starting to gain some traction in C++, with STL classes like `std::sort()` that take comparison functors. Those usually get declared anonymously right before they're passed in to the library class.

Reply:
> however I'm not sure about the placement of the function. Unless it is used by something else than `findAllMarkerPositions()`, I think it is cleanest to leave it inside. 

That's my thinking, as you know.

> And otherwise, maybe finding another place than `main_window.py`, if it is not inside the `MainWindow` class

Because it deals actively — _inter_actively, in fact — with selection and UI navigation, it's not really appropriate for any of the `src/classes/` modules where helper functions typically reside. So it has to live _somewhere_ in the `src/windows/` tree. The only other reasonable place would be somewhere in `src/windows/views/timeline_webview.py`.

But IMHO it's already right where it belongs.

Reply:
@ferdnyc I see no arguments in your words but only feelings. It slower. It is harder to read. So, what benefits of "nesting" here?

Reply:
@SuslikV I agree with @ferdnyc. 
Performance is the only truly objective metric here, and yes, nesting the function is a slight performance loss, but it's objectively negligible. (number of additional cpu cycles would be orders of magnitude lower than total) 
Code readability however, is subjective but important. And I think it's much much easier to read with the function being nested than anywhere else. It would make no sense to me to have the main_window file start with a function that has no obvious role for the mainwindow. In the calling method however, it makes sense. From that method, we know exactly what the relevance of the callee is for the caller, it's close by in the code so no scrolling up and down far, and as the nested function is never called by anything else than it's parent method, so there is no use or sense exposing it to a larger scope.

If performance was the only important aspect, we would be using a purely imperative language, with no objects, no functions... they are all convenient overheads that sacrifice performance for code clarity. I think we should make the same choice here.

Reply:
I would actually be ok with changing `getTimelineObjectPositions` to a method of MainWindow, and declare it just above `findAllMarkerPositions`. Less nesting so better performance and geographically close to the caller so it's easy to read. But it's still uselessly exposed to the MainWindow scope, and it's also unnecessary for it to be a method, it does not need any access to the instance

Reply:
> If performance was the only important aspect, we would be using a purely imperative language, with no objects, no functions... 

This. If we were being miserly with every CPU cycle, for starters we'd _never_ write `try`/`except` blocks — the penalty for handling even a _single_ exception is **orders of magnitude** higher than the alternative ugly forest of conditionals it could be replaced with. But even though standard conditionals are much, _much_ faster, programmers accept the performance penalty for exception handling. As long as you avoid writing tight loops that have to handle thousands of exceptions per second, the additional cost is a relatively small price to pay.

>they are all convenient overheads that sacrifice performance for code clarity.

I wouldn't even say "sacrifice", since we're talking about a function that's only called in response to an explicit user action. The limiting factor on performance in those cases is reaction time — first the interface's, then the _user's_. As long as the UI remains fully responsive, performance is at its effective maximum level.

I mean, during each second you only get 60 chances to update the position of the playhead. Within each of those 60ths of a second, things can't possibly happen any faster no matter how many cycles you shave off the algorithm.

@SuslikV 
> I see no arguments in your words but only feelings.

They're not "feelings", they're opinions. Just like this is an opinion:

> Nested functions is hard to read. 

Not only do I disagree, I don't really understand the objection. Every class method is already a nested function, effectively. People somehow manage.

The simple fact is, I personally find local worker functions to be _more_readable than classes stuffed with utility methods that aren't really part of their API. The function is right there in the method where it's used, and it's only exposed to the one piece of code that's meant to be calling it. So we'll have to agree to disagree on this point, since readability is an opinion.

Reply:
@ferdnyc So, why it is here?!! WHY??? For what purpose nesting is needed here? For what? I don't understand it!

Reply:
@ferdnyc , It was wrong suggestion to use nesting... and you know it.